src/core/Queue.cpp

```diff
@@ -193,87 +196,103 @@ bool Queue::isEmpty() const
   return m_queue.empty();
 }
 
-Queue::Command* Queue::update()
+void Queue::execute(Command *command)
 {
-  if (m_queue.empty())
+  // Find command in queue
+  auto it = std::find(m_queue.begin(), m_queue.end(), command);
+
+  // If this queue is not out of order, add previous event associated with
+  // previous command in queue as dependency
+  if (it != m_queue.begin() /* && Check if not out of order queue */)
   {
-    return NULL;
+    command->waitList.push_back((*std::prev(it))->event);
   }
 
-  // Get next command
-  Command *cmd = m_queue.front();
-
-  // Check if all events in wait list have completed
-  while (!cmd->waitList.empty())
+  // Make sure all events in the wait list are complete before executing
+  // current command
+  while (!command->waitList.empty())
   {
-    if (cmd->waitList.front()->state == CL_COMPLETE)
-    {
-      cmd->waitList.pop_front();
-    }
-    else if (cmd->waitList.front()->state < 0)
+    Event *evt = command->waitList.front();
+    command->waitList.pop_front();
+
+    if (evt->state < 0)
     {
-      cmd->event->state = cmd->waitList.front()->state;
-      m_queue.pop();
-      return cmd;
+      command->event->state = evt->state;
+      return;
     }
-    else
+    else if (evt->state != CL_COMPLETE)
     {
-      return NULL;
+      evt->queue->execute(evt->command);
```

Comment:
Might need to special case for user events here, since there won't be a command associated with it (and we probably just need to spinlock until it is complete). The OpenCL conformance test suite has decent coverage for this sort of thing in the `events` category, which is worth running if you're able to (we used to pass all tests in that category).

Reply:
Yes, this was one thing I wasn't 100% confident. I'll have a go at it tomorrow and check with the OpenCL CTS.

Reply:
Update:
* The `master` branch is failing almost half of the CTS event tests (branch `cl12_trunk`). 
* Additionally, and as you suspected, this PR is causing the `userevents` test to fail (besides the ones already failing in `master`). The issue is slightly more complicated than expected, but I think I've nailed it down and will probably submit a fix tomorrow.

Reply:
Commit 8c1e5ce1def573df96c3bbb6b30cede2bd773122 fixes the failing `userevents` test.

Note that the following CTS event tests are failing (in both `master` and this branch):

            //test_event_get_execute_status,
            //test_event_get_info,
            //test_event_wait_for_execute,
            //test_event_flush,
            //test_event_finish_execute,
            //test_event_release_before_done,
            //test_event_enqueue_marker,
            //test_event_enqueue_marker_with_list,
            //test_event_enqueue_barrier_with_list,
            //test_callbacks,
            //test_callbacks_simultaneous,
            //test_userevents_multithreaded,

I might have a look at them in the future, as some of these might impact my use cases.

Reply:
**Update**

So, I decided to get to the bottom of this, and it seems the tests were not hanging, just taking a lot of time. Reducing the loop iterations in the test kernels allows the tests to finish in a reasonable time. As such, all CTS event tests are passing in both `master` and this `multiple_queues` branch.